File: sdk/typescript/src/namespaces/nomic.ts
/**
 * Nomic Namespace API
 *
 * Provides a namespaced interface for nomic loop control and monitoring.
 * The nomic loop is the autonomous self-improvement cycle that enables
 * the system to propose, design, implement, and verify its own improvements.
 */

/**
 * Nomic loop health status
 */
export type NomicHealthStatus = 'healthy' | 'stalled' | 'not_running' | 'error';

/**
 * Nomic loop phase
 */
export type NomicPhase = 'context' | 'debate' | 'design' | 'implement' | 'verify';

/**
 * Nomic loop state
 */
export interface NomicState {
  running: boolean;
  cycle: number;
  phase: NomicPhase;
  paused: boolean;
  started_at?: string;
  last_update?: string;
  target_cycles?: number;
  auto_approve?: boolean;
}

/**
 * Nomic loop health
 */
export interface NomicHealth {
  status: NomicHealthStatus;
  cycle: number;
  phase: NomicPhase | null;
  last_activity: string | null;
  stall_duration_seconds: number | null;
  warnings: string[];
}

/**
 * Nomic loop metrics summary
 */
export interface NomicMetrics {
  summary: Record<string, unknown>;
  stuck_detection: {
    is_stuck: boolean;
    phase?: string;
    idle_seconds?: number;
  };
  status: string;
}

/**
 * Nomic log response
 */
export interface NomicLogResponse {
  lines: string[];
  total: number;
  showing: number;
}

/**
 * Risk register entry
 */
export interface RiskEntry {
  id: string;
  severity: 'critical' | 'high' | 'medium' | 'low';
  description: string;
  category: string;
  created_at: string;
  resolved_at?: string;
  metadata?: Record<string, unknown>;
}

/**
 * Risk register response
 */
export interface RiskRegisterResponse {
  risks: RiskEntry[];
  total: number;
  critical_count: number;
  high_count: number;
}

/**
 * Improvement proposal
 */
export interface Proposal {
  id: string;
  title: string;
  description: string;
  status: 'pending' | 'approved' | 'rejected';
  submitted_at: string;
  approved_at?: string;
  rejected_at?: string;
  approved_by?: string;
  rejected_by?: string;
  rejection_reason?: string;
  changes?: Record<string, unknown>;
}

/**
 * Proposals list response
 */
export interface ProposalsResponse {
  proposals: Proposal[];
  total: number;
  all_proposals: number;
}

/**
 * Start loop request
 */
export interface StartLoopRequest {
  cycles?: number;
  max_cycles?: number;
  auto_approve?: boolean;
  dry_run?: boolean;
}

/**
 * Stop loop request
 */
export interface StopLoopRequest {
  graceful?: boolean;
}

/**
 * Approve proposal request
 */
export interface ApproveProposalRequest {
  proposal_id: string;
  approved_by?: string;
}

/**
 * Reject proposal request
 */
export interface RejectProposalRequest {
  proposal_id: string;
  rejected_by?: string;
  reason?: string;
}

/**
 * Operational mode
 */
export interface OperationalMode {
  name: string;
  type: 'builtin' | 'custom';
  description: string;
  slug?: string;
}

/**
 * Gas Town witness status
 */
export interface WitnessStatus {
  active: boolean;
  patrol_count: number;
  last_patrol: string | null;
  violations_detected: number;
  current_focus?: string;
  witnesses: Array<{
    id: string;
    name: string;
    status: 'active' | 'idle' | 'investigating';
    last_report: string | null;
  }>;
}

/**
 * Gas Town mayor information
 */
export interface MayorInfo {
  current_mayor: string | null;
  elected_at: string | null;
  term_ends: string | null;
  approval_rating: number | null;
  policies_enacted: number;
  vetoes: number;
  emergency_powers_active: boolean;
}

/**
 * Modes response
 */
export interface ModesResponse {
  modes: OperationalMode[];
  total: number;
}

/**
 * Interface for the internal client used by NomicAPI.
 */
interface NomicClientInterface {
  get<T>(path: string): Promise<T>;
  post<T>(path: string, body?: unknown): Promise<T>;
  request<T>(method: string, path: string, options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }): Promise<T>;
}

/**
 * Nomic API namespace.
 *
 * Provides methods for nomic loop control and monitoring:
 * - Get loop state, health, and metrics
 * - Start, stop, pause, and resume the loop
 * - Manage improvement proposals
 * - Access risk register and logs
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai', apiKey: 'your-key' });
 *
 * // Check nomic loop health
 * const health = await client.nomic.getHealth();
 * console.log(`Status: ${health.status}, Phase: ${health.phase}`);
 *
 * // Start nomic loop
 * await client.nomic.start({ cycles: 3, auto_approve: false });
 *
 * // Get pending proposals
 * const { proposals } = await client.nomic.getProposals();
 *
 * // Approve a proposal
 * await client.nomic.approveProposal({ proposal_id: 'prop-123' });
 *
 * // Get risk register
 * const risks = await client.nomic.getRiskRegister({ limit: 50 });
 * ```
 */
export class NomicAPI {
  constructor(private client: NomicClientInterface) {}

  // ===========================================================================
  // State & Monitoring
  // ===========================================================================

  /**
   * Get current nomic loop state.
   */
  async getState(): Promise<NomicState> {
    return this.client.get('/api/v1/nomic/state');
  }

  /**
   * Get nomic loop health with stall detection.
   */
  async getHealth(): Promise<NomicHealth> {
    return this.client.get('/api/v1/nomic/health');
  }

  /**
   * Get nomic loop metrics summary.
   */
  async getMetrics(): Promise<NomicMetrics> {
    return this.client.get('/api/v1/nomic/metrics');
  }

  /**
   * Get recent nomic loop log lines.
   * @param lines - Number of log lines to return (default: 100, max: 1000)
   */
  async getLogs(lines: number = 100): Promise<NomicLogResponse> {
    return this.client.request('GET', '/api/v1/nomic/log', { params: { lines } });
  }

  /**
   * Get risk register entries.
   * @param options - Filter options
   */
  async getRiskRegister(options?: { limit?: number }): Promise<RiskRegisterResponse> {
    return this.client.request('GET', '/api/v1/nomic/risk-register', { params: options });
  }

  // ===========================================================================
  // Loop Control
  // ===========================================================================

  /**
   * Start the nomic loop.
   * @param body - Start configuration
   */
  async start(body?: StartLoopRequest): Promise<{ status: string; pid: number; target_cycles: number }> {
    return this.client.post('/api/v1/nomic/control/start', body);
  }

  /**
   * Stop the running nomic loop.
   * @param body - Stop configuration
   */
  async stop(body?: StopLoopRequest): Promise<{ status: string; pid: number }> {
    return this.client.post('/api/v1/nomic/control/stop', body);
  }

  /**
   * Pause the nomic loop at the current phase.
   */
  async pause(): Promise<{ status: string; cycle: number; phase: string }> {
    return this.client.post('/api/v1/nomic/control/pause');
  }

  /**
   * Resume a paused nomic loop.
   */
  async resume(): Promise<{ status: string; cycle: number; phase: string }> {
    return this.client.post('/api/v1/nomic/control/resume');
  }

  /**
   * Skip the current phase and move to the next.
   */
  async skipPhase(): Promise<{ status: string; previous_phase: string; next_phase: string; cycle: number }> {
    return this.client.post('/api/v1/nomic/control/skip-phase');
  }

  // ===========================================================================
  // Proposal Management
  // ===========================================================================

  /**
   * Get pending improvement proposals.
   */
  async getProposals(): Promise<ProposalsResponse> {
    return this.client.get('/api/v1/nomic/proposals');
  }

  /**
   * Approve a pending proposal.
   * @param body - Approval details
   */
  async approveProposal(body: ApproveProposalRequest): Promise<{ status: string; proposal_id: string }> {
    return this.client.post('/api/v1/nomic/proposals/approve', body);
  }

  /**
   * Reject a pending proposal.
   * @param body - Rejection details
   */
  async rejectProposal(body: RejectProposalRequest): Promise<{ status: string; proposal_id: string }> {
    return this.client.post('/api/v1/nomic/proposals/reject', body);
  }

  // ===========================================================================
  // Operational Modes
  // ===========================================================================

  /**
   * Get available operational modes (builtin + custom).
   */
  async getModes(): Promise<ModesResponse> {
    return this.client.get('/api/v1/modes');
  }

  // ===========================================================================
  // Gas Town
  // ===========================================================================

  /**
   * Get Gas Town witness patrol status.
   *
   * Witnesses monitor the nomic loop for violations and irregularities.
   */
  async getWitnessStatus(): Promise<WitnessStatus> {
    return this.client.get('/api/v1/nomic/witness/status');
  }

  /**
   * Get current Gas Town mayor information.
   *
   * The mayor is the elected leader who can enact policies and use emergency powers.
   */
  async getMayorCurrent(): Promise<MayorInfo> {
    return this.client.get('/api/v1/nomic/mayor/current');
  }

  // ===========================================================================
  // Convenience Aliases
  // ===========================================================================

  /**
   * Alias for getState().
   */
  async state(): Promise<NomicState> {
    return this.getState();
  }

  /**
   * Alias for getHealth().
   */
  async health(): Promise<NomicHealth> {
    return this.getHealth();
  }

  /**
   * Check if the nomic loop is currently running.
   */
  async isRunning(): Promise<boolean> {
    const state = await this.getState();
    return state.running && !state.paused;
  }

  /**
   * Get proposal by ID from the proposals list.
   *
   * @param proposalId - The proposal ID to find
   */
  async getProposal(proposalId: string): Promise<Proposal | null> {
    const { proposals } = await this.getProposals();
    return proposals.find(p => p.id === proposalId) || null;
  }
}

File: sdk/python/aragora/namespaces/nomic.py
"""
Nomic Namespace API

Provides methods for nomic loop control and monitoring:
- Get loop state, health, and metrics
- Start, stop, pause, and resume the loop
- Manage improvement proposals
- Access risk register and logs
- Operational modes management
- Gas Town witness and mayor monitoring

The nomic loop is the autonomous self-improvement cycle that enables
the system to propose, design, implement, and verify its own improvements.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class NomicAPI:
    """
    Synchronous Nomic API.

    Provides methods for nomic loop control and monitoring:
    - Get loop state, health, and metrics
    - Start, stop, pause, and resume the loop
    - Manage improvement proposals
    - Access risk register and logs
    - Operational modes management
    - Gas Town witness and mayor monitoring

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> health = client.nomic.get_health()
        >>> print(f"Status: {health['status']}, Phase: {health['phase']}")
        >>> client.nomic.start(cycles=3, auto_approve=False)
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    # ===========================================================================
    # State & Monitoring
    # ===========================================================================

    def get_state(self) -> dict[str, Any]:
        """
        Get current nomic loop state.

        Returns:
            Dict with running, cycle, phase, paused, started_at, last_update,
            target_cycles, and auto_approve fields
        """
        return self._client.request("GET", "/api/v1/nomic/state")

    def get_health(self) -> dict[str, Any]:
        """
        Get nomic loop health with stall detection.

        Returns:
            Dict with status (healthy/stalled/not_running/error), cycle, phase,
            last_activity, stall_duration_seconds, and warnings
        """
        return self._client.request("GET", "/api/v1/nomic/health")

    def get_metrics(self) -> dict[str, Any]:
        """
        Get nomic loop metrics summary.

        Returns:
            Dict with summary, stuck_detection, and status
        """
        return self._client.request("GET", "/api/v1/nomic/metrics")

    def get_logs(self, lines: int = 100) -> dict[str, Any]:
        """
        Get recent nomic loop log lines.

        Args:
            lines: Number of log lines to return (default: 100, max: 1000)

        Returns:
            Dict with lines array, total, and showing count
        """
        return self._client.request("GET", "/api/v1/nomic/log", params={"lines": lines})

    def get_risk_register(self, limit: int | None = None) -> dict[str, Any]:
        """
        Get risk register entries.

        Args:
            limit: Maximum entries to return

        Returns:
            Dict with risks array, total, critical_count, and high_count
        """
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        return self._client.request("GET", "/api/v1/nomic/risk-register", params=params)

    # ===========================================================================
    # Loop Control
    # ===========================================================================

    def start(
        self,
        cycles: int | None = None,
        max_cycles: int | None = None,
        auto_approve: bool = False,
        dry_run: bool = False,
    ) -> dict[str, Any]:
        """
        Start the nomic loop.

        Args:
            cycles: Target number of cycles to run
            max_cycles: Alias for cycles (deprecated)
            auto_approve: Auto-approve proposals without human review
            dry_run: Preview changes without executing

        Returns:
            Dict with status, pid, and target_cycles
        """
        data: dict[str, Any] = {
            "auto_approve": auto_approve,
            "dry_run": dry_run,
        }
        if cycles is not None:
            data["cycles"] = cycles
        elif max_cycles is not None:
            data["max_cycles"] = max_cycles
        return self._client.request("POST", "/api/v1/nomic/control/start", json=data)

    def stop(self, graceful: bool = True) -> dict[str, Any]:
        """
        Stop the running nomic loop.

        Args:
            graceful: Allow current phase to complete (default: True)

        Returns:
            Dict with status and pid
        """
        data: dict[str, Any] = {"graceful": graceful}
        return self._client.request("POST", "/api/v1/nomic/control/stop", json=data)

    def pause(self) -> dict[str, Any]:
        """
        Pause the nomic loop at the current phase.

        Returns:
            Dict with status, cycle, and phase
        """
        return self._client.request("POST", "/api/v1/nomic/control/pause")

    def resume(self) -> dict[str, Any]:
        """
        Resume a paused nomic loop.

        Returns:
            Dict with status, cycle, and phase
        """
        return self._client.request("POST", "/api/v1/nomic/control/resume")

    def skip_phase(self) -> dict[str, Any]:
        """
        Skip the current phase and move to the next.

        Returns:
            Dict with status, previous_phase, next_phase, and cycle
        """
        return self._client.request("POST", "/api/v1/nomic/control/skip-phase")

    # ===========================================================================
    # Proposal Management
    # ===========================================================================

    def get_proposals(self) -> dict[str, Any]:
        """
        Get pending improvement proposals.

        Returns:
            Dict with proposals array, total, and all_proposals count
        """
        return self._client.request("GET", "/api/v1/nomic/proposals")

    def approve_proposal(
        self,
        proposal_id: str,
        approved_by: str | None = None,
    ) -> dict[str, Any]:
        """
        Approve a pending proposal.

        Args:
            proposal_id: The proposal ID to approve
            approved_by: Identifier of the approver

        Returns:
            Dict with status and proposal_id
        """
        data: dict[str, Any] = {"proposal_id": proposal_id}
        if approved_by:
            data["approved_by"] = approved_by
        return self._client.request("POST", "/api/v1/nomic/proposals/approve", json=data)

    def reject_proposal(
        self,
        proposal_id: str,
        rejected_by: str | None = None,
        reason: str | None = None,
    ) -> dict[str, Any]:
        """
        Reject a pending proposal.

        Args:
            proposal_id: The proposal ID to reject
            rejected_by: Identifier of the rejector
            reason: Reason for rejection

        Returns:
            Dict with status and proposal_id
        """
        data: dict[str, Any] = {"proposal_id": proposal_id}
        if rejected_by:
            data["rejected_by"] = rejected_by
        if reason:
            data["reason"] = reason
        return self._client.request("POST", "/api/v1/nomic/proposals/reject", json=data)

    # ===========================================================================
    # Operational Modes
    # ===========================================================================

    def get_modes(self) -> dict[str, Any]:
        """
        Get available operational modes (builtin + custom).

        Returns:
            Dict with modes array and total count
        """
        return self._client.request("GET", "/api/v1/modes")

    # ===========================================================================
    # Gas Town (Witness & Mayor)
    # ===========================================================================

    def get_witness_status(self) -> dict[str, Any]:
        """
        Get Gas Town witness patrol status.

        Witnesses monitor the nomic loop for violations and irregularities.

        Returns:
            Dict with active, patrol_count, last_patrol, violations_detected,
            current_focus, and witnesses array
        """
        return self._client.request("GET", "/api/v1/nomic/witness/status")

    def get_mayor_current(self) -> dict[str, Any]:
        """
        Get current Gas Town mayor information.

        The mayor is the elected leader who can enact policies and use emergency powers.

        Returns:
            Dict with current_mayor, elected_at, term_ends, approval_rating,
            policies_enacted, vetoes, and emergency_powers_active
        """
        return self._client.request("GET", "/api/v1/nomic/mayor/current")

    # ===========================================================================
    # Convenience Methods
    # ===========================================================================

    def state(self) -> dict[str, Any]:
        """Alias for get_state()."""
        return self.get_state()

    def health(self) -> dict[str, Any]:
        """Alias for get_health()."""
        return self.get_health()

    def is_running(self) -> bool:
        """
        Check if the nomic loop is currently running.

        Returns:
            True if running and not paused
        """
        state = self.get_state()
        return state.get("running", False) and not state.get("paused", False)

    def get_proposal(self, proposal_id: str) -> dict[str, Any] | None:
        """
        Get a specific proposal by ID.

        Args:
            proposal_id: The proposal ID to find

        Returns:
            Proposal dict or None if not found
        """
        response = self.get_proposals()
        for proposal in response.get("proposals", []):
            if proposal.get("id") == proposal_id:
                return proposal
        return None


class AsyncNomicAPI:
    """
    Asynchronous Nomic API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     health = await client.nomic.get_health()
        ...     await client.nomic.start(cycles=3, auto_approve=False)
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    # ===========================================================================
    # State & Monitoring
    # ===========================================================================

    async def get_state(self) -> dict[str, Any]:
        """Get current nomic loop state."""
        return await self._client.request("GET", "/api/v1/nomic/state")

    async def get_health(self) -> dict[str, Any]:
        """Get nomic loop health with stall detection."""
        return await self._client.request("GET", "/api/v1/nomic/health")

    async def get_metrics(self) -> dict[str, Any]:
        """Get nomic loop metrics summary."""
        return await self._client.request("GET", "/api/v1/nomic/metrics")

    async def get_logs(self, lines: int = 100) -> dict[str, Any]:
        """Get recent nomic loop log lines."""
        return await self._client.request("GET", "/api/v1/nomic/log", params={"lines": lines})

    async def get_risk_register(self, limit: int | None = None) -> dict[str, Any]:
        """Get risk register entries."""
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        return await self._client.request("GET", "/api/v1/nomic/risk-register", params=params)

    # ===========================================================================
    # Loop Control
    # ===========================================================================

    async def start(
        self,
        cycles: int | None = None,
        max_cycles: int | None = None,
        auto_approve: bool = False,
        dry_run: bool = False,
    ) -> dict[str, Any]:
        """Start the nomic loop."""
        data: dict[str, Any] = {
            "auto_approve": auto_approve,
            "dry_run": dry_run,
        }
        if cycles is not None:
            data["cycles"] = cycles
        elif max_cycles is not None:
            data["max_cycles"] = max_cycles
        return await self._client.request("POST", "/api/v1/nomic/control/start", json=data)

    async def stop(self, graceful: bool = True) -> dict[str, Any]:
        """Stop the running nomic loop."""
        data: dict[str, Any] = {"graceful": graceful}
        return await self._client.request("POST", "/api/v1/nomic/control/stop", json=data)

    async def pause(self) -> dict[str, Any]:
        """Pause the nomic loop at the current phase."""
        return await self._client.request("POST", "/api/v1/nomic/control/pause")

    async def resume(self) -> dict[str, Any]:
        """Resume a paused nomic loop."""
        return await self._client.request("POST", "/api/v1/nomic/control/resume")

    async def skip_phase(self) -> dict[str, Any]:
        """Skip the current phase and move to the next."""
        return await self._client.request("POST", "/api/v1/nomic/control/skip-phase")

    # ===========================================================================
    # Proposal Management
    # ===========================================================================

    async def get_proposals(self) -> dict[str, Any]:
        """Get pending improvement proposals."""
        return await self._client.request("GET", "/api/v1/nomic/proposals")

    async def approve_proposal(
        self,
        proposal_id: str,
        approved_by: str | None = None,
    ) -> dict[str, Any]:
        """Approve a pending proposal."""
        data: dict[str, Any] = {"proposal_id": proposal_id}
        if approved_by:
            data["approved_by"] = approved_by
        return await self._client.request("POST", "/api/v1/nomic/proposals/approve", json=data)

    async def reject_proposal(
        self,
        proposal_id: str,
        rejected_by: str | None = None,
        reason: str | None = None,
    ) -> dict[str, Any]:
        """Reject a pending proposal."""
        data: dict[str, Any] = {"proposal_id": proposal_id}
        if rejected_by:
            data["rejected_by"] = rejected_by
        if reason:
            data["reason"] = reason
        return await self._client.request("POST", "/api/v1/nomic/proposals/reject", json=data)

    # ===========================================================================
    # Operational Modes
    # ===========================================================================

    async def get_modes(self) -> dict[str, Any]:
        """Get available operational modes."""
        return await self._client.request("GET", "/api/v1/modes")

    # ===========================================================================
    # Gas Town (Witness & Mayor)
    # ===========================================================================

    async def get_witness_status(self) -> dict[str, Any]:
        """Get Gas Town witness patrol status."""
        return await self._client.request("GET", "/api/v1/nomic/witness/status")

    async def get_mayor_current(self) -> dict[str, Any]:
        """Get current Gas Town mayor information."""
        return await self._client.request("GET", "/api/v1/nomic/mayor/current")

    # ===========================================================================
    # Convenience Methods
    # ===========================================================================

    async def state(self) -> dict[str, Any]:
        """Alias for get_state()."""
        return await self.get_state()

    async def health(self) -> dict[str, Any]:
        """Alias for get_health()."""
        return await self.get_health()

    async def is_running(self) -> bool:
        """Check if the nomic loop is currently running."""
        state = await self.get_state()
        return state.get("running", False) and not state.get("paused", False)

    async def get_proposal(self, proposal_id: str) -> dict[str, Any] | None:
        """Get a specific proposal by ID."""
        response = await self.get_proposals()
        for proposal in response.get("proposals", []):
            if proposal.get("id") == proposal_id:
                return proposal
        return None
